### PR TITLE
!!! [v6r17] Fix JobScheduling optimizer

### PR DESCRIPTION
**Pull request by phicharp**
* JobScheduling is calling getFilesToStage that toggles UseServerCertificate. As the optimizer is multithreaded this must be protected with a lock.
* getProxy has a bad indentation preventing looping on the DNs
* Minor other fixes

### CODE REVIEW

**Comment by fstagni** [comment]
The unit test fails most probably because the "mockDM" used is not up-to-date (i.e. does not include the `getReplicasForJobs` mocked function). The file is https://github.com/DIRACGrid/DIRAC/blob/integration/StorageManagementSystem/Client/test/Test_Client_StorageManagementSystem.py

You may modify the `mockObjectDM` in that file and adding the `return_value` also for `getReplicasForJobs`, or (better option), update and use the generic mocked DM that is found at https://github.com/DIRACGrid/DIRAC/blob/integration/DataManagementSystem/Client/test/mock_DM.py and in which we should try and collect all the other instances of mocks of the DataManager object.

**Comment by phicharp** [comment]
Will look at it later... Right, mock needs to be updated...

**Comment by phicharp** [comment]
I don't understand how it could be the problem as I only replace getActiveReplicas with getReplicasForJobs... and getActiveReplicas was NOT int the mock... In addition this test has nothing to do AFAIS with JobScheduling... Let's see what comes out of the test this time

**Comment by fstagni** [comment]
The file https://github.com/DIRACGrid/DIRAC/blob/integration/StorageManagementSystem/Client/test/Test_Client_StorageManagementSystem.py is NOT, as of now, using the `mock_DM.py` file that you just modified: so, just modify the Test_Client_StorageManagementSystem.py file to use the mock_DM file.

**Comment by phicharp** [comment]
These methods may return different dictionaries depending on the other arguments (e.g. `getUrl`). How can we properly test that it is OK to not return a URL for example?

**Comment by atsareg** [comment]
For the thread locking  it should be LockRing object used rather than the raw threading module lock.

**Comment by chaen** [comment]
Why ? Can you explain a bit what is this LockRing object, I never really knew

**Comment by atsareg** [comment]
This is a DIRAC utility singleton global object that holds named locks. It behaves like a threading lock with some added features. In particular, it is reset in ProcessPool when creating new processes to avoid dead locks like we had in the past with multithreaded agents.

**Comment by atsareg** [comment]
By the way, does the threading lock in this PR actually solve the problem ? If I understand correctly, the lock is created for the JobScheduling executor only. But the executor itself is not multithreaded, rather different executors are running in parallel threads. So, this should be a global object for all the executors. Am I missing something ?

**Comment by chaen** [comment]
Ok, but why in this particular case should we use this ? We don't have multi process, so there is no deadlock. Or do I miss something ?
The lock is indeed for the JobScheduling only, but all of the jobschedulling instances, since it is a class attribute. There are no other executors changing the UseServerCertificate

**Comment by atsareg** [comment]
There is no strict necessity to use LockRing. With #3250 you do not need even to create a lock object.

InputData executor potentially can also have the same problem. I think LHCb does not set the flag to use it with user proxy though. But to keep the code generic then the InputData operations should be also called with a lock if executed with user proxy. 

As for the JobScheduling - there is only one thread ( per Optimizers process ), so how can it step on itself ?

**Comment by phicharp** [comment]
A bit too technical for me ;-)
But the optimizers are not using the DIRAC threads, they just use the threading module AFAIS, so why use a DIRAC lock?
As I have no clue how executors work, I don't understand how one executor can be single threaded and the other one multithreaded... Anyway, we have put printing at the place where we switch UseServerCertificate and it happens we reach it while it has the wrong value. The timing also shows that the method is started while the previous call is still executing, which shows IMHO that there is more than one thread calling it...

**Comment by phicharp** [comment]
Added a minor fix when setting files Processed un UploadOutputData

**Comment by phicharp** [comment]
Using the executionLock option of #3250 
In getFilesToStage, ignore files that are on disk

Please re-review!

File: StorageManagementSystem/Client/StorageManagerClient.py [hunk 1]
```diff
@@ -23,51 +23,64 @@ def getFilesToStage( lfnList ):
...
+      gLogger.error( "Could not get metadata", "for %d files" % len( set( [lfn for lfnList in failed.itervalues() for lfn in lfnList] ) ) )
+      return S_ERROR( "Could not get metadata for files" )
+    offlineLFNs = set( lfnList ) - onlineLFNs
 
-  for offlineLFN in offlineLFNs:
-    ses = lfnListReplicas['Value']['Successful'][offlineLFN].keys()
-    if ses:
-      offlineLFNsDict.setdefault( random.choice( ses ), list() ).append( offlineLFN )
+    for offlineLFN in offlineLFNs:
+      ses = lfnListReplicas['Value']['Successful'][offlineLFN].keys()
```
Comment (atsareg) [inline StorageManagementSystem/Client/StorageManagerClient.py]: ses = lfnListReplicas['Value']['Successful'][offlineLFN].keys() -> ses = lfnListReplicas[offlineLFN].keys()

**Comment by atsareg** [comment]
I have made #3251 ( to try ) to make Travis happy

File: StorageManagementSystem/Client/StorageManagerClient.py (hunk 1 above)
Comment (phicharp) [inline StorageManagementSystem/Client/StorageManagerClient.py]: oups, sorry, change at one place and not the other one, well seen!

**Pull request merged**